### PR TITLE
feat: improve lineage perf

### PR DESCRIPTION
**Pull request by tobymao**
this should significantly increase the perf of lineage, will require sqlglot 23.1

### CODE REVIEW

**Review by vchan — APPROVED** [review]
Should we trim selects here as well? https://github.com/TobikoData/sqlmesh/blob/main/web/server/api/endpoints/lineage.py#L56

**Comment by tobymao** [comment]
> Should we trim selects here as well? https://github.com/TobikoData/sqlmesh/blob/main/web/server/api/endpoints/lineage.py#L56

no, we shouldn't. but this breaks some tests so i'll need some time to fix it and test

File: sqlmesh/core/lineage.py [hunk 1]
```diff
@@ -15,44 +16,63 @@
...
-            *[
-                exp.alias_(exp.cast(exp.Null(), to=col_type), col)
-                for col, col_type in model.managed_columns.items()
-                if col not in query.named_selects
-            ],
-            append=True,
-            copy=False,
-        )
-    return query
+CACHE: t.Dict[str, t.Tuple[int, exp.Expression, Scope]] = {}
```
Comment (georgesittas) [inline sqlmesh/core/lineage.py:19]: Should this be global? Can it ever be modified concurrently & if so, would there be issues?

**Review by vchan — APPROVED** [review]

File: web/server/api/endpoints/lineage.py [hunk 2]
```diff
@@ -115,16 +115,10 @@ async def column_lineage(
...
-        return create_lineage_adjacency_list(model_name, column_name, context)
-    except Exception:
-        raise ApiException(
-            message="Unable to get column lineage",
-            origin="API -> lineage -> column_lineage",
-        )
+    model_name = context.get_model(model_name).fqn
+    if models_only:
+        return create_models_only_lineage_adjacency_list(model_name, column_name, context)
+    return create_lineage_adjacency_list(model_name, column_name, context)
```
Comment (georgesittas) [inline web/server/api/endpoints/lineage.py:121]: We don't need to raise the `ApiException` here anymore?

File: sqlmesh/core/lineage.py [hunk 3]
```diff
@@ -15,44 +16,63 @@
...
+    if not query or not scope:
+        query = model.render_query_or_raise()
+
+        if model.managed_columns:
+            query = query.select(
+                *(
+                    exp.alias_(exp.cast(exp.Null(), to=col_type), col)
+                    for col, col_type in model.managed_columns.items()
+                    if col not in query.named_selects
+                )
```
Comment (georgesittas) [inline sqlmesh/core/lineage.py:47]: Should we set `copy=False` here?

**Review by georgesittas — APPROVED** [review]
Looks good, only some questions FMI.

File: sqlmesh/core/lineage.py (hunk 1 above)
Comment (tobymao) [reply sqlmesh/core/lineage.py:19]: in general our code is not really thread safe, so this is fine i think

File: sqlmesh/core/lineage.py (hunk 3 above)
Comment (tobymao) [reply sqlmesh/core/lineage.py:47]: no, we don't want to do mutate query

**Pull request merged**